### PR TITLE
TimezoneDropdown: get timezones from WP REST-API

### PR DESCRIPTION
**Pull request by retrofox**
In this PR the `TimezoneDropdown` is filled with the values gotten from `/timezones` endpoint.
I've removed timezone.js (moment.js) in favor to be consistent between WP.com and calypso.

In this way if the timezones defined in WP.com change the response of `/timezones` will change too keeping the same data structure.

Related issue: #3902

### Testing

1) Go to `TimezoneDropdown` component testing page and play switching the timezones. 
http://calypso.localhost:3000/devdocs/design/timezone-dropdown

2) Open the dev console. Everything should be ok.

image
1

### CODE REVIEW

**Comment by artpi** [comment]
I tested it and `master` works:


But this branch - not so much:


I know this is still work in progress, just letting know :)

File: client/components/timezone-dropdown/docs/example.jsx [hunk 1]
```diff
@@ -34,7 +34,7 @@ export default React.createClass( {
 					<a href="/devdocs/design/timezone-dropdown">TimezoneDropdown</a>
 				</h2>
 
-				<Card style={ { width: '300px', margin: 0 } }>
+				<Card style={ { width: '300px', height: '350px', margin: 0 } }>
```
Comment (artpi) [inline client/components/timezone-dropdown/docs/example.jsx]: I am not sure this is the solution here.
Probably dropdown needs some Z-index magic, because if we use it in settings, we wont be making vertical space for it, we want it to go over the other settings.

File: client/components/timezone-dropdown/index.jsx [hunk 2]
```diff
@@ -3,50 +3,79 @@
...
+ */
+const { Component, PropTypes } = React;
+const noop = () => {};
+const undocumented = wpcom.undocumented();
+
+class TimezoneDropdown extends Component {
+	constructor() {
+		super();
 
-	propTypes: {
```
Comment (artpi) [inline client/components/timezone-dropdown/index.jsx]: When you'll be finishing this, please remember to put propTypes here!

Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: Why should we put `propTypes` here? They are defined below.

Comment (artpi) [inline client/components/timezone-dropdown/index.jsx]: I see now! I was not expecting that! :D

File: client/components/timezone-dropdown/docs/example.jsx (hunk 1 above)
Comment (retrofox) [inline client/components/timezone-dropdown/docs/example.jsx]: I think it should be handled directly in devdocs because the component is mounted pretty well.

image

**Comment by retrofox** [comment]
Added _Manual Offsets_

image

**Comment by rralian** [comment]
Honestly I find this UX very difficult to use. It's very difficult for me to quickly skim through the list to get to where I need. It's too easy to skip through `America` for example and drop into the next category without realizing it.

In wp-admin we use a native select element which works fine because it expands to the size of my screen which helps me to easily see where I am. Whereas this component doesn't open as much, I can't see as much at once. I think the indentation of item values relative to the label category in a native select element also helps to keep your place as you quickly skim through the options, even in mobile where you don't see as much at once. I understand why the suggestion was made to keep everything left-aligned in the prior PR, but in practice this makes for a difficult UX with a giant list like this that requires a fast flick through options. One other thing... the highlighting of selected item versus highlighting of a hovered item adds a lot of visual complexity and further makes it difficult to distinguish the labels.

I would prefer that we combine the city/continent into a single value, like: `America/Chicago` and skip the labels completely, OR that we indent the option values relative to the category labels, OR that we provide a search box interface, so I can quickly type in "chicago" and find what I'm looking for versus having to hunt and peck. Though that last suggestion is a really big change.

/cc @adambbecker for his thoughts.

### Difficult to differentiate label from item
screen shot 2016-03-23 at 3 53 51 pm

### Missing context of category here
screen shot 2016-03-23 at 3 42 26 pm

### Native select element in mobile has indentation
img_9731

### Native select in desktop (chrome at least) shows many more items
screen shot 2016-03-23 at 3 41 07 pm

**Comment by rralian** [comment]
> Added Manual Offsets

I think the manual offsets can all just live under the `UTC` category. I would start with `UTC` as the first item and then show the negative items `UTC-12` and progress up through the positive values.

File: client/components/select-dropdown/index.jsx [hunk 3]
```diff
@@ -43,8 +43,7 @@ class SelectDropdown extends Component {
 		let initialState = { isOpen: false };
```
Comment (gwwar) [inline client/components/select-dropdown/index.jsx]: Any reason why we not using the standard `getInitialState`?

File: client/components/timezone-dropdown/index.jsx [hunk 4]
```diff
@@ -3,50 +3,100 @@
...
-		selectedZone: React.PropTypes.string,
-		onSelect: React.PropTypes.func
-	},
+		// bound methods
+		this.onSelect = this.onSelect.bind( this );
```
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: Why do you need to bind onSelect? Class functions should already have proper scope.

File: client/components/timezone-dropdown/index.jsx [hunk 5]
```diff
@@ -3,50 +3,100 @@
...
-	getTimezoneNames() {
-		return i18n.moment.tz.names().map( zone => {
-			return ( {
-				label: zone,
-				value: zone
-			} );
+	}
+
+	componentWillMount() {
+		undocumented.timezones( ( err, zones ) => {
```
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: This is a good candidate to put into redux

File: client/components/timezone-dropdown/index.jsx [hunk 6]
```diff
@@ -3,50 +3,100 @@
...
+			if ( err ) {
+				return console.log( err );
```
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: How are we handling errors visually?

File: client/components/timezone-dropdown/index.jsx [hunk 7]
```diff
@@ -3,50 +3,100 @@
...
+};
+
+TimezoneDropdown.defaultProps = {
+	onSelect: noop
+};
+
+TimezoneDropdown.propTypes = {
+	selectedZone: PropTypes.string,
+	onSelect: PropTypes.func
+}
```
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: Why are these defined down here, out of the class?

**Comment by mtias** [comment]
I think labels could remain fixed at the top while you scroll until you get to the next label:

image

**Comment by retrofox** [comment]
> > Added Manual Offsets

> I think the manual offsets can all just live under the UTC category. I would start with UTC as the first item and then show the negative items UTC-12 and progress up through the positive values.

It's fine for me. I just followed the wp-admin options.

image

**Comment by retrofox** [comment]
> I think labels could remain fixed at the top while you scroll until you get to the next label:

That would be great. :+1:

File: client/components/timezone-dropdown/index.jsx (hunk 7 above)
Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: I've followed this doc:

https://facebook.github.io/react/docs/reusable-components.html#es6-classes

> Another difference is that propTypes and defaultProps are defined as properties on the constructor instead of in the class body.

File: client/components/timezone-dropdown/index.jsx (hunk 6 above)
Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: oh, I forgot improve this handling ... I think I'll just silence it.

File: client/components/timezone-dropdown/index.jsx (hunk 5 above)
Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: :-o can we do it in another PR please? 
I promise to make this improvement as soon as it is merged.

File: client/components/timezone-dropdown/docs/example.jsx (hunk 1 above)
Comment (rralian) [inline client/components/timezone-dropdown/docs/example.jsx]: I don't think this is specific to this instance of the dropdown component, but it seems they do not expand beyond the `id="content"` node of the page. I think that's likely not a common issue, but one we should be careful about. I don't think we need to address it in this PR though.
screen shot 2016-03-23 at 4 27 04 pm

File: client/components/timezone-dropdown/index.jsx (hunk 2 above)
Comment (rralian) [inline client/components/timezone-dropdown/index.jsx]: Do we have a preference for this pattern? I also hadn't seen it before in our codebase, but maybe I'm just missing it.

File: client/components/timezone-dropdown/index.jsx (hunk 4 above)
Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: yes ..., but `onSelect` is called from a component property and there the function isn't bounded and on other hand we should avoid call a bound method from the react prop.

```es6
<TimezoneDropdown
  onClick={ this.onSelect.bind( this ) }
/>
```

File: client/components/timezone-dropdown/index.jsx (hunk 5 above)
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: :+1:

File: client/components/timezone-dropdown/index.jsx (hunk 7 above)
Comment (gwwar) [inline client/components/timezone-dropdown/index.jsx]: :+1: got it. Missed the plain es6 class

**Comment by gwwar** [comment]
:+1: Works as advertised. A suggested future improvement may be to add typing support. I normally do this on native selects to quickly navigate to a particular item.

File: client/components/timezone-dropdown/index.jsx (hunk 2 above)
Comment (retrofox) [inline client/components/timezone-dropdown/index.jsx]: https://facebook.github.io/react/docs/reusable-components.html#es6-classes

> ```es6
export class Counter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {count: props.initialCount};
  }
  tick() {
    this.setState({count: this.state.count + 1});
  }
  render() {
    return (
      <div onClick={this.tick.bind(this)}>
        Clicks: {this.state.count}
      </div>
    );
  }
}
Counter.propTypes = { initialCount: React.PropTypes.number };
Counter.defaultProps = { initialCount: 0 };
```

Also I've seen this pattern in some code written by @aduth. am I right?

**Comment by retrofox** [comment]
I've added padding to timezones strings. Let's keep the following improvements for others PRs:

* remain labels fixed at the top while you scroll until you get to the next label
* focus the selected option when the a timezone is already selected
* add search input field ?

**Comment by retrofox** [comment]
Final version in this PR:

image

**Comment by rralian** [comment]
> I've added padding to timezones strings. Let's keep the following improvements for others PRs:

Agreed. Let's launch this. :boom:

File: client/components/timezone-dropdown/docs/example.jsx (hunk 1 above)
Comment (retrofox) [inline client/components/timezone-dropdown/docs/example.jsx]: :+1:
